### PR TITLE
iOS Simulator and iOS platform support

### LINKED ISSUE #627: How to compile and run on iOS?

### PR DESCRIPTION
**Pull request by gpx1000**
## Description

Add support for iOS

Fixes #627 

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

**Comment by gpx1000** [comment]
NB: this fails currently on the VolkInitialize().

File: docs/build.md [hunk 1]
```diff
@@ -269,6 +269,57 @@ cmake --build build/mac --config Release --target vulkan_samples -j4
...
+
+```
+cmake -H. -Bbuild/ios -DCMAKE_BUILD_TYPE=Release -G Xcode -DCMAKE_SYSTEM_NAME=iOS -DCMAKE_OSX_DEPLOYMENT_TARGET=11.0
+```
+
+`Step 2.` Build the project
+
+```
+cmake --build build/ios --config Release --target 
+vulkan_samples -j4 -sdk iphonesimulator -allowProvisioningUpdates
```
Comment (tomadamatkinson) [inline docs/build.md]: Does this need to be on a single line?

File: bldsys/cmake/global_options.cmake [hunk 2]
```diff
@@ -32,12 +32,9 @@ else()
 endif()
 
 if(APPLE)
-    # attempt to use Find_Vulkan at least of version 1.3 or greater.  If this is found, then enable the required portability extension so instance.cpp can get created correctly.
-    # prior to 1.3 Find_Vulkan should not return a Vulkan_FOUND variable, so should not set an extension that doesn't exist in legacy Vulkan.
-    # Note that this is only required for moltenVK implementations.  NB: this does create a warning in CMake that the range isn't supported in FindVulkan.
-    cmake_minimum_required(VERSION 3.19)
-    find_package(Vulkan 1.3...<2.0)
-    if(Vulkan_FOUND)
+    cmake_minimum_required(VERSION 3.24)
```
Comment (tomadamatkinson) [inline bldsys/cmake/global_options.cmake:35]: Can we require a CMake version smaller than 3.24?

Comment (tomadamatkinson) [reply bldsys/cmake/global_options.cmake:35]: I have updated now from 3.22 to 3.27 on my Macbook. But it is always best to use the minimum version. Happy to keep this change if 3.24 is the new required version

File: docs/build.md [hunk 3]
```diff
@@ -269,6 +269,57 @@ cmake --build build/mac --config Release --target vulkan_samples -j4
...
+```
+
+NB: The above creates a bundle.  It is possible to build
```
Comment (tomadamatkinson) [inline docs/build.md]: Is it worth adding Xcode instructions?

For me this was the following with my iphone plugged in

On the Iphone
1. Go the Settings app, and navigate to Privacy & Security > Developer Mode
2. Restart when prompted

On the macbook
1. Open Xcode -> Open the Xcode project from build/ios
3. Set the target to vulkan_samples
4. Follow code signing steps https://developer.apple.com/documentation/xcode/running-your-app-in-simulator-or-on-a-device
5. Set the device to a simulator (default) or the plugged in iphone
6. Click Play

We could also make an issue to follow up with these docs and do it at a later date as these would likely require screenshot examples

File: framework/platform/ios/ios_window.mm [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+void IosWindow::close()
+{
+	//	GameActivity_finish(platform->get_activity());
+	finish_called = true;
+}
+
+float IosWindow::get_dpi_factor() const
+{
+	//	return AConfiguration_getDensity(platform->get_android_app()->config) / static_cast<float>(ACONFIGURATION_DENSITY_MEDIUM);
```
Comment (tomadamatkinson) [inline framework/platform/ios/ios_window.mm]: remove unused code

File: framework/platform/ios/ios_window.mm [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+bool IosWindow::should_close()
+{
+	return finish_called;
+}
+
+void IosWindow::close()
+{
+	//	GameActivity_finish(platform->get_activity());
```
Comment (tomadamatkinson) [inline framework/platform/ios/ios_window.mm]: remove unused code

File: framework/platform/ios/ios_window.mm [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	info.pLayer = (CAMetalLayer *) [window layer];
+
+	VK_CHECK(vkCreateMetalSurfaceEXT(instance, &info, nullptr, &surface));
+
+	return surface;
+}
+
+void IosWindow::process_events()
+{
+	//	process_android_events(platform->get_android_app());
```
Comment (tomadamatkinson) [inline framework/platform/ios/ios_window.mm]: remove unused code

File: app/CMakeLists.txt [hunk 7]
```diff
@@ -80,4 +80,19 @@ if(MSVC)
...
-endif()
\ No newline at end of file
+endif()
+
+# Enable building from XCode for IOS
+if(IOS)
+    set(CMAKE_MACOSX_BUNDLE YES)
+    set(MACOSX_BUNDLE_GUI_IDENTIFIER com.khronos.vulkan_samples)
+    set_target_properties(${PROJECT_NAME} PROPERTIES
+            BUNDLE_IDENTIFIER com.khronos.vulkan_samples
```
Comment (tomadamatkinson) [inline app/CMakeLists.txt]: Code signing doesn't work with `com.khronos.vulkan_samples` but did work with `com.khronos.VulkanSamples`

File: docs/build.md (hunk 3 above)
Comment (tomadamatkinson) [reply docs/build.md]: There are a lot of code signing steps but I eventually got the application to run

**Review by tomadamatkinson — CHANGES_REQUESTED** [review]
The application starts but Vulkan initialization fails on the device and in the simulator. Did you need to perform any extra steps to get Vulkan to work?

File: bldsys/cmake/global_options.cmake (hunk 2 above)
Comment (gpx1000) [reply bldsys/cmake/global_options.cmake:35]: TBH, with the new VulkanSDK coming out where we use xcframework, we're going to need to bump it to 3.28  But we can work around that if we need to.

File: docs/build.md (hunk 3 above)
Comment (gpx1000) [reply docs/build.md]: I have now gotten this to work without XCode needing to be open.  However, now, you can simply open the project in XCode and hitting the play button; it'll simply work no need for the extra work.

**Comment by richard-lunarg** [comment]
I've tried this locally. As of today, it builds with the current macOS VulkanSDK, and the upcoming 1.3.280.1 release.

**Review by richard-lunarg — APPROVED** [review]
I've tried this PR locally. It builds with the latest cmake and Xcode versions for me and I can deploy to an iOS device(s).

**Comment by jeroenbakker-atmind** [comment]
I tested compiling this PR with VulkanSDK 1.3.275.0 to ensure old workflows still work. Validated both the CMake and Xcode path. Both worked as expected. I don't have an iOS (or simulator) currently installed, but that has already been validated by others.

**Pull request merged**